client_request_metrics_test.py

```diff
@@ -0,0 +1,784 @@
+                              SimpleStatement(f"SELECT k FROM {KEYSPACE}.{TABLE} WHERE k=0 AND c=0", consistency_level=CL.LOCAL_ONE),
+                              query_count)
+
+    def validate_nominal(self, global_scope, metric_class, statement, query_count):
+        query_cl = statement.consistency_level
+        cl_scope = scope_for_cl(global_scope, query_cl)
+        # Validate other CLs metrics are not changing. We're not messing with Quorum because of the async write coming from
+        # role creation at startup.
+        # If the test takes long enough, it's possible to see some of the time-based metrics shift.
+        other_cls = [c for c in CL.value_to_name if c not in (query_cl, CL.QUORUM)]
+        global_baseline = metric_class(global_scope)
+        cl_baseline = metric_class(cl_scope)
+        other_baselines = [metric_class(scope_for_cl(global_scope, c)) for c in other_cls]
+
+
+        global_baseline.validate()
+        cl_baseline.validate()
+        for b in other_baselines:
+            b.validate()
+
+        for _ in range(query_count):
+            self.session.execute(statement)
+
+        global_updated = metric_class(global_scope)
+        cl_updated = metric_class(cl_scope)
+        global_updated.validate()
+        cl_updated.validate()
+
+        global_diff = global_updated.diff(global_baseline)
+        cl_diff = cl_updated.diff(cl_baseline)
+        for diff in [global_diff, cl_diff]:
+            assert diff['TotalLatency.Count'] > 0, diff.scope
+            assert diff['Latency.Count'] == query_count, diff.scope
+
+        other_updated = [metric_class(scope_for_cl(global_scope, c)) for c in other_cls]
+        for updated, baseline in zip(other_updated, other_baselines):
+            assert 'Latency.Count' not in updated.diff(baseline), updated.scope
+
+        return global_diff, cl_diff
+
+    def write_failures(self):
+        query_count = 2
+        diff = self.write_failures_variant('Write', 'WHERE k=0 AND c=0',
+                                           query_count,
+                                           self.validate_exception_metric_with_cl,
+                                           WriteMetrics)
+        assert diff['MutationSizeHistogram.Count'] == query_count
+
+    def write_failures_variant(self, scope, constraint, query_count, validator, metric_class):
+        validator = self.validate_metric if validator is None else validator
+        query_cl = CL.ONE
+        diff = validator(scope,
+                         metric_class,
+                         SimpleStatement(
+                             f"UPDATE {FAIL_WRITE_KEYSPACE}.{TABLE} SET v=0 {constraint}",
+                             consistency_level=query_cl),
+                         query_count,
+                         'Failures',
+                         WriteFailure
+                         )
+        return diff
+
+    def write_unavailables(self):
+        query_cl = CL.THREE
```

Comment:
We could add a short comment indicating that the writes are unavailable because the replication factor is 2.